Source/SwiftLintBuiltInRules/Rules/Lint/TodoRule.swift

```diff
@@ -32,54 +25,90 @@ struct TodoRule: ConfigurationProviderRule {
         ].skipWrappingInCommentTests()
     )
 
-    private func customMessage(file: SwiftLintFile, range: NSRange) -> String {
-        var reason = Self.description.description
-        let offset = NSMaxRange(range)
+    func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
+    }
+}
 
-        guard let (lineNumber, _) = file.stringView.lineAndCharacter(forCharacterOffset: offset) else {
-            return reason
+private extension TodoRule {
+    final class Visitor: ViolationsSyntaxVisitor {
+        override func visitPost(_ node: TokenSyntax) {
+            let leadingViolations = node.leadingTrivia.violations(offset: node.position)
+            let trailingViolations = node.trailingTrivia.violations(offset: node.endPositionBeforeTrailingTrivia)
+            violations.append(contentsOf: leadingViolations + trailingViolations)
         }
+    }
+}
 
-        let line = file.lines[lineNumber - 1]
-        // customizing the reason message to be specific to fixme or todo
-        let violationSubstring = file.stringView.substring(with: range)
+private extension Trivia {
+    func violations(offset: AbsolutePosition) -> [ReasonedRuleViolation] {
+        var position = offset
+        var violations = [ReasonedRuleViolation]()
+        for piece in self {
+            violations.append(contentsOf: piece.violations(offset: position))
+            position = position.advanced(by: piece.sourceLength.utf8Length)
+        }
+        return violations
+    }
+}
 
-        let range = NSRange(location: offset, length: NSMaxRange(line.range) - offset)
-        var message = file.stringView.substring(with: range)
-        let kind = violationSubstring.hasPrefix("FIXME") ? "FIXMEs" : "TODOs"
+private extension TriviaPiece {
+    func violations(offset: AbsolutePosition) -> [ReasonedRuleViolation] {
+        switch self {
+        case
+                .blockComment(let comment),
+                .lineComment(let comment),
+                .docBlockComment(let comment),
+                .docLineComment(let comment):
+            var startIndex = comment.startIndex
+            var violations = [ReasonedRuleViolation]()
+            while let result = try? NSRegularExpression.cached(pattern: #"\b(TODO|FIXME)(?::|\b)([^\n]*)"#)
+                .firstMatch(in: comment, range: NSRange(startIndex..., in: comment)) {
```

Comment:
Out of interest: Can we use `matches` here instead of `firstMatch` to avoid the `while` loop?

Reply:
Probably.